tokio/src/task/task_local.rs

```diff
@@ -224,76 +301,109 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
-        future: F,
-        #[pin]
-        _pinned: PhantomPinned,
-    }
+/// A future that sets a value `T` of a task local for the future `F` during
+/// its execution.
+///
+/// The value of the task-local must be `'static` and will be dropped on the
+/// completion of the future.
+///
+/// Created by the function [`LocalKey::scope`](self::LocalKey::scope).
+///
+/// ### Examples
+///
+/// ```
+/// # async fn dox() {
+/// tokio::task_local! {
+///     static NUMBER: u32;
+/// }
+///
+/// NUMBER.scope(1, async move {
+///     println!("task local value: {}", NUMBER.get());
+/// }).await;
+/// # }
+/// ```
+// Doesn't use pin_project due to custom Drop.
+pub struct TaskLocalFuture<T, F>
+where
+    T: 'static,
+{
+    local: &'static LocalKey<T>,
+    slot: Option<T>,
+    future: Option<F>,
+    _pinned: PhantomPinned,
 }
 
-impl<T: 'static, F> TaskLocalFuture<T, F> {
-    fn with_task<F2: FnOnce(Pin<&mut F>) -> R, R>(self: Pin<&mut Self>, f: F2) -> R {
-        struct Guard<'a, T: 'static> {
-            local: &'static LocalKey<T>,
-            slot: &'a mut Option<T>,
-            prev: Option<T>,
-        }
-
-        impl<T> Drop for Guard<'_, T> {
-            fn drop(&mut self) {
-                let value = self.local.inner.with(|c| c.replace(self.prev.take()));
-                *self.slot = value;
-            }
-        }
+impl<T: 'static, F: Future> Future for TaskLocalFuture<T, F> {
+    type Output = F::Output;
 
-        let project = self.project();
-        let val = project.slot.take();
+    #[track_caller]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // safety: The TaskLocalFuture struct is `!Unpin` so there is no way to
+        // move `self.future` from now on.
+        let this = unsafe { Pin::into_inner_unchecked(self) };
+        let mut future_opt = unsafe { Pin::new_unchecked(&mut this.future) };
 
-        let prev = project.local.inner.with(|c| c.replace(val));
+        let res =
+            this.local
+                .scope_inner(&mut this.slot, || match future_opt.as_mut().as_pin_mut() {
+                    Some(fut) => {
+                        let res = fut.poll(cx);
+                        if res.is_ready() {
+                            future_opt.set(None);
+                        }
+                        Some(res)
+                    }
+                    None => None,
+                });
 
-        let _guard = Guard {
-            prev,
-            slot: project.slot,
-            local: *project.local,
-        };
+        match res {
+            Ok(Some(res)) => res,
+            Ok(None) => panic!("TaskLocalFuture polled after completion"),
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("TaskLocalFuture::poll called while task local is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
```

Comment:
tiny nit: this panic message shows up in a couple places, maybe it should be a const?